storage/pebble/operation/common.go

```diff
@@ -34,17 +93,342 @@ func retrieve(key []byte, sc interface{}) func(r pebble.Reader) error {
+		// for discussion and more detail on this.
+
+		// If start is bigger than end, we have a backwards iteration:
+		// 1) We set the reverse option on the iterator, so we step through all
+		//    the keys backwards. This modifies the behaviour of Seek to go to
+		//    the first key that is less than or equal to the start key (as
+		//    opposed to greater than or equal in a regular iteration).
+		// 2) In order to satisfy this function's prefix-wise inclusion semantics,
+		//    we append a 0xff-byte suffix to the start key so the seek will go
+		// to the right place.
+		// 3) For a regular iteration, we break the loop upon hitting the first
+		//    item that has a key higher than the end prefix. In order to reverse
+		//    this, we use a modifier for the comparison that reverses the check
+		//    and makes it stop upon the first item lower than the end prefix.
+		// for forward iteration, add the 0xff-bytes suffix to the end
+		// prefix, to ensure we include all keys with that prefix before
+		// finishing.
+		it, err := tx.NewIter(&options)
+		if err != nil {
+			return fmt.Errorf("can not create iterator: %w", err)
+		}
+		defer it.Close()
+
+		for it.SeekGE(start); it.Valid(); it.Next() {
+			key := it.Key()
+			// Break the loop if we have passed the end key prefix
+			if bytes.Compare(key, end) > 0 && !startsWithPrefix(key, end) {
+				break
+			}
+
+			// initialize processing functions for iteration
+			check, create, handle := iteration()
+
+			// check if we should process the item at all
+			ok := check(key)
+			if !ok {
+				continue
+			}
+
+			// when prefetchValues is false, we skip loading the value
+			if !prefetchValues {
+				continue
+			}
+
+			binaryValue, err := it.ValueAndErr()
+			if err != nil {
+				return fmt.Errorf("failed to get value: %w", err)
+			}
+
+			// preventing caller from modifying the iterator's value slices
+			valueCopy := make([]byte, len(binaryValue))
+			copy(valueCopy, binaryValue)
+
+			entity := create()
+			err = msgpack.Unmarshal(valueCopy, entity)
+			if err != nil {
+				return irrecoverable.NewExceptionf("could not decode entity: %w", err)
+			}
+
+			// process the entity
+			err = handle()
+			if err != nil {
+				return fmt.Errorf("could not handle entity: %w", err)
+			}
+		}
+
+		return nil
+	}
+}
+
+var ffBytes = bytes.Repeat([]byte{0xFF}, 32)
+
+// traverse iterates over a range of keys defined by a prefix.
+//
+// The prefix must be shared by all keys in the iteration.
+//
+// On each iteration, it will call the iteration function to initialize
+// functions specific to processing the given key-value pair.
+func traverse(prefix []byte, iteration iterationFunc) func(pebble.Reader) error {
+	return func(r pebble.Reader) error {
+		if len(prefix) == 0 {
+			return fmt.Errorf("prefix must not be empty")
+		}
+
+		it, err := r.NewIter(&pebble.IterOptions{
+			LowerBound: prefix,
+			UpperBound: append(prefix, ffBytes...),
```

Comment:
The badger implementation maintains a global `max` key, so that when iterating from a key range with a start prefix and end prefix, it uses the max key to[ create the boundary for including the biggest key with the end prefix. ](https://github.com/onflow/flow-go/blob/leo/v0.33-pebble-storage-to-review/storage/badger/operation/common.go#L405-L410): 
```
		} else {
			// for forward iteration, add the 0xff-bytes suffix to the end
			// prefix, to ensure we include all keys with that prefix before
			// finishing.
			length := uint32(len(end))
			diff := max - length
			for i := uint32(0); i < diff; i++ {
				end = append(end, 0xff)
			}
		}
```

I found a way to simplify it by getting rid of this max value completely.

I noticed when creating the key range, we are usually just iterating identifiers, such as iterating the child block IDs, or iterating the event IDs of a block, or iterating transaction results of a block. They are either flow.Identifier or uint32, which takes 32 bytes. So I could just make a ffBytes appended with 32 FF bytes, and use it as the UpperBound for the key range bounded by the `end` prefix.

Reply:
☠️ see [my comment blow for details](https://github.com/onflow/flow-go/pull/6197#issuecomment-2278752293). Unfortunately, I think that change is unacceptable.

Reply:
Agree with the worries about this change. The `0xff` suffix approach to prefix-wise iteration depends on tracking database-wide key lengths to be correct. The Badger solution was a little hacky, but at least provided a fairly strong guarantee of correctness by forcing all writes through logic that tracked the max key length. Adding a constant number of bytes is less likely to be robust and correct over time because it does not account for changes to key lengths.

We have encountered very subtle bugs caused by not handling this correctly in the past. (The link on line 356 is supposed to link to [this explanation of such a bug](https://github.com/dapperlabs/flow-go/pull/3310#issuecomment-618127494)).

### Change Suggestions

Pebble has an [example for prefix-wise iteration](https://pkg.go.dev/github.com/cockroachdb/pebble#example-Iterator-PrefixIteration) in their documentation -- maybe we can use that approach here instead?

Alternatively, we could explicitly overshoot the endpoint of the iteration range, and exit early if we seek to a key with a prefix that doesn't match the supplied prefix.

Reply:
Thanks for the ideas, I made the changes [in this PR](https://github.com/onflow/flow-go/pull/6316) in order to highlight the changes and the updated test cases. 

>it's probably easier and more accurate to omit upper bound and check if the iterator's current key starts with prefix.

This could work for `traverse`, but won't work for `iterate` which iterates keys between a start prefix and a end prefix. We would have to implement two solutions, whereas we could just use one solution for both `traverse` and `iterate`.